backend/telemetry_core/src/main.rs

```diff
@@ -0,0 +1,468 @@
+                        req,
+                        move |ws_send, ws_recv| async move {
+                            log::info!("Opening /shard_submit connection from {:?}", addr);
+                            let tx_to_aggregator = aggregator.subscribe_shard();
+                            let (mut tx_to_aggregator, mut ws_send) =
+                                handle_shard_websocket_connection(
+                                    ws_send,
+                                    ws_recv,
+                                    tx_to_aggregator,
+                                )
+                                .await;
+                            log::info!("Closing /shard_submit connection from {:?}", addr);
+                            // Tell the aggregator that this connection has closed, so it can tidy up.
+                            let _ = tx_to_aggregator
+                                .send(FromShardWebsocket::Disconnected)
+                                .await;
+                            let _ = ws_send.close().await;
+                        },
+                    ))
+                }
+                // 404 for anything else:
+                _ => Ok(Response::builder()
+                    .status(404)
+                    .body("Not found".into())
+                    .unwrap()),
+            }
+        }
+    });
+
+    server.await?;
+    Ok(())
+}
+
+/// This handles messages coming to/from a shard connection
+async fn handle_shard_websocket_connection<S>(
+    mut ws_send: http_utils::WsSender,
+    mut ws_recv: http_utils::WsReceiver,
+    mut tx_to_aggregator: S,
+) -> (S, http_utils::WsSender)
+where
+    S: futures::Sink<FromShardWebsocket, Error = anyhow::Error> + Unpin + Send + 'static,
+{
+    let (tx_to_shard_conn, mut rx_from_aggregator) = mpsc::unbounded();
+
+    // Tell the aggregator about this new connection, and give it a way to send messages to us:
+    let init_msg = FromShardWebsocket::Initialize {
+        channel: tx_to_shard_conn,
+    };
+    if let Err(e) = tx_to_aggregator.send(init_msg).await {
+        log::error!("Error sending message to aggregator: {}", e);
+        return (tx_to_aggregator, ws_send);
+    }
+
+    // Channels to notify each loop if the other closes:
+    let (recv_closer_tx, mut recv_closer_rx) = tokio::sync::oneshot::channel::<()>();
+    let (send_closer_tx, mut send_closer_rx) = tokio::sync::oneshot::channel::<()>();
+
+    // Receive messages from a shard:
+    let recv_handle = tokio::spawn(async move {
+        loop {
+            let mut bytes = Vec::new();
+
+            // Receive a message, or bail if closer called. We don't care about cancel safety;
+            // if we're halfway through receiving a message, no biggie since we're closing the
+            // connection anyway.
+            let msg_info = tokio::select! {
+                msg_info = ws_recv.receive_data(&mut bytes) => msg_info,
+                _ = &mut recv_closer_rx => { break }
```

Comment:
style grumble, I don't like the braces :P

```suggestion
                _ = &mut recv_closer_rx => break,
```